compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -480,15 +480,18 @@ impl<'a> Linker for GccLinker<'a> {
     fn control_flow_guard(&mut self) {}
 
     fn debuginfo(&mut self, strip: Strip) {
+        // MacOS linker doesn't support stripping symbols directly anymore.
+        if self.sess.target.is_like_osx {
+            return;
+        }
```

Comment:
Wouldn't this preclude people from using this functionality when using alternative linkers, if available? I'm thinking of lld, which I understand is broken, but there's no inherent reason for it, right?

Reply:
that's a good point. I'm open to suggestions on how to make this check more reliable across linkers.

Reply:
is there any other precedence of different behavior between linkers?

Reply:
There's nothing to say some alternative (as yet unknown) linker would support `--strip-debug` and `--strip-all` either though, so aren't we just getting into the realms of speculation?

Reply:
I suppose we could add a configuration option that determines which path to take?  That pushes the responsibility to decide out of rustc and onto the invoker (Cargo?).  Perhaps that's where this whole thing belongs anyway.

Reply:
I just want to toss out one option is to add a new linker flavor for darwin (or macos or ld64 or whatever it should be called).  Although the "common" flags are pretty much the same between darwin and gcc, many specialized flags are different.  The code could reuse `GccLinker`, and just add another flag similar to `is_ld` (or store the flavor itself).

I'm not sure if this has been discussed before. I feel like it has, but I can't find any discussions.

Reply:
> I just want to toss out one option is to add a new linker flavor for darwin

I have the feeling that your comment is deviating from the point of this PR, but I might be missing some information. The provided linker that macos uses doesn't have the options to strip symbols. You have to use the `strip` command after the linker runs to strip the symbols. It's true that other linkers might have the option, and people could solve it by using a different linker, but the problem with the default linker would still exist. I think providing a different linker should be the goal of a different change.

Reply:
I cannot tell from @calavera's response whether their interpretation of @ehuss's feedback is the same as my own interpretation of it.

My interpretation of the feedback is that @ehuss is suggesting that instead of putting hard-coded logic in this spot that inspect's `target.is_like_osx`, instead you could add a new variant to the `enum LinkerFlavor` here: 

https://github.com/rust-lang/rust/blob/df81891dbd849ab2a8893aa472a1a9f1f6d52ed0/compiler/rustc_target/src/spec/mod.rs#L88-L95

That is, it is not that @ehuss is saying that your change should be trying to provide a totally different linker. They are instead saying the linker should stay the same, but the way that you *specify* how this linker differs from the other linkers is by encoding the distinction in the `enum` (and then have that enum drive the decisions about what specific command-line options to generate, I think).

Reply:
that makes a lot more sense. Thanks for the explanation @pnkfelix